### PR TITLE
HTML Search: Add test for querying

### PR DESCRIPTION
**Pull request by wlach**
Subject: Add a unit test for querying functionality in search

### Feature or Bugfix

- Bugfix / Tests

### Purpose

The search query function is currently untested, though it's a key piece of functionality.

Intended to eventually be updated after #11942 is landed (but seperating out for now for ease-of-review).

### Detail

On investigation, there were a number of minor bugs in search term functionality, which were fixed along with this PR.

### Relates
- #11942

### CODE REVIEW

File: sphinx/themes/basic/static/searchtools.js [hunk 1]
```diff
@@ -367,8 +372,7 @@ const Search = {
     //Search.lastresults = results.slice();  // a copy
     // console.info("search results:", Search.lastresults);
 
-    // print the results
-    _displayNextItem(results, results.length, searchTerms, highlightTerms);
+    return { results, searchTerms, highlightTerms };
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js:375]: Refactored this to be seperate from displaying the results, which was necessary to add the test.

File: sphinx/themes/basic/static/searchtools.js [hunk 2]
```diff
@@ -463,17 +467,23 @@ const Search = {
         { files: terms[word], score: Scorer.term },
         { files: titleTerms[word], score: Scorer.title },
       ];
+
       // add support for partial matches
       if (word.length > 2) {
         const escapedWord = _escapeRegExp(word);
-        Object.keys(terms).forEach((term) => {
-          if (term.match(escapedWord) && !terms[word])
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js:470]: `terms[word]` is a problem, since if it it's referring to the 0th item, it will have a false value.

This is probably rare in production unless you happen to be unlucky, but can easily happen in a test (see below)

File: sphinx/themes/basic/static/searchtools.js [hunk 3]
```diff
@@ -463,17 +467,23 @@ const Search = {
...
-            arr.push({ files: terms[term], score: Scorer.partialTerm });
-        });
-        Object.keys(titleTerms).forEach((term) => {
-          if (term.match(escapedWord) && !titleTerms[word])
-            arr.push({ files: titleTerms[word], score: Scorer.partialTitle });
-        });
+        if (!terms.hasOwnProperty(word)) {
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js:474]: Checking for hasOwnProperty before iterating over all the terms seems like an easy optimization, so I just did it.

File: sphinx/themes/basic/static/searchtools.js [hunk 4]
```diff
@@ -496,9 +506,13 @@ const Search = {
 
       // create the mapping
       files.forEach((file) => {
-        if (fileMap.has(file) && fileMap.get(file).indexOf(word) === -1)
```
Comment (wlach) [inline sphinx/themes/basic/static/searchtools.js:499]: Another fun little bug -- if the file was already present, but the word wasn't there, it would immediately jump to the else clause to reinitialize the map from scratch. Another edge case that would be more likely to happen in a test.

File: tests/js/searchtools.js [hunk 5]
```diff
@@ -21,14 +21,41 @@ describe('Basic html theme search', function() {
         "&lt;no title&gt;",
         "",
         null,
-        2,
+        5,
```
Comment (wlach) [inline tests/js/searchtools.js:24]: The score was artificially low in this test because we were overwriting the main match with a partial match due to this bug:

https://github.com/sphinx-doc/sphinx/pull/11950/files#r1478405652

File: tests/js/searchtools.js [hunk 6]
```diff
@@ -21,14 +21,41 @@ describe('Basic html theme search', function() {
...
         "index.rst"
       ]];
       expect(Search.performTermsSearch(searchterms, excluded, terms, titleterms)).toEqual(hits);
     });
 
   });
 
+  describe('query', function() {
+    it("should not duplicate on title and content", function() {
```
Comment (wlach) [inline tests/js/searchtools.js:33]:
```suggestion
    it("search where we match both title and content", function() {
```

File: tests/js/searchtools.js [hunk 7]
```diff
@@ -21,14 +21,41 @@ describe('Basic html theme search', function() {
...
+        5,
         "index.rst"
       ]];
       expect(Search.performTermsSearch(searchterms, excluded, terms, titleterms)).toEqual(hits);
     });
 
   });
 
+  describe('query', function() {
+    it("should duplicate on title and content", function() {
```
Comment (wlach) [inline tests/js/searchtools.js:33]: Change this to "should not" later after #11942 lands

**Comment by wlach** [comment]
@jayaddison Another search ticket to look at, probably useful in the context of #11942 and #11944 (no rush obviously)

**Comment by wlach** [comment]
Moved this out into other PRs, closing.
